<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Pairwise Distances from DNA Sequences</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for dist.dna {ape}"><tr><td>dist.dna {ape}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Pairwise Distances from DNA Sequences</h2>

<h3>Description</h3>

<p>This function computes a matrix of pairwise distances from DNA
sequences using a model of DNA evolution. Eleven substitution models
(and the raw distance) are currently available.
</p>


<h3>Usage</h3>

<pre>
dist.dna(x, model = "K80", variance = FALSE,
         gamma = FALSE, pairwise.deletion = FALSE,
         base.freq = NULL, as.matrix = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>a matrix or a list containing the DNA sequences; this must be
of class <code>"DNAbin"</code> (use <code><a href="as.alignment.html">as.DNAbin</a></code> is they are
stored as character).</p>
</td></tr>
<tr valign="top"><td><code>model</code></td>
<td>
<p>a character string specifying the evolutionary model to be
used; must be one of <code>"raw"</code>, <code>"N"</code>, <code>"TS"</code>,
<code>"TV"</code>, <code>"JC69"</code>, <code>"K80"</code> (the default),
<code>"F81"</code>, <code>"K81"</code>, <code>"F84"</code>, <code>"BH87"</code>,
<code>"T92"</code>, <code>"TN93"</code>, <code>"GG95"</code>, <code>"logdet"</code>,
<code>"paralin"</code>, <code>"indel"</code>, or <code>"indelblock"</code>.</p>
</td></tr>
<tr valign="top"><td><code>variance</code></td>
<td>
<p>a logical indicating whether to compute the variances
of the distances; defaults to <code>FALSE</code> so the variances are not
computed.</p>
</td></tr>
<tr valign="top"><td><code>gamma</code></td>
<td>
<p>a value for the gamma parameter possibly used to apply a
correction to the distances (by default no correction is applied).</p>
</td></tr>
<tr valign="top"><td><code>pairwise.deletion</code></td>
<td>
<p>a logical indicating whether to delete the
sites with missing data in a pairwise way. The default is to delete
the sites with at least one missing data for all sequences (ignored
if <code>model = "indel"</code> or <code>"indelblock"</code>).</p>
</td></tr>
<tr valign="top"><td><code>base.freq</code></td>
<td>
<p>the base frequencies to be used in the computations
(if applicable). By default, the base frequencies are computed from
the whole set of sequences.</p>
</td></tr>
<tr valign="top"><td><code>as.matrix</code></td>
<td>
<p>a logical indicating whether to return the results as
a matrix. The default is to return an object of class
<a href="../../stats/html/dist.html">dist</a>.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The molecular evolutionary models available through the option
<code>model</code> have been extensively described in the literature. A
brief description is given below; more details can be found in the
references.
</p>

<ul>
<li><p><code>raw</code>, <code>N</code>: This is simply the proportion or the number of
sites that differ between each pair of sequences. This may be useful
to draw &ldquo;saturation plots&rdquo;. The options <code>variance</code> and
<code>gamma</code> have no effect, but <code>pairwise.deletion</code> can.
</p>
</li>
<li><p><code>TS</code>, <code>TV</code>: These are the numbers of transitions and
transversions, respectively.
</p>
</li>
<li><p><code>JC69</code>: This model was developed by Jukes and Cantor (1969). It
assumes that all substitutions (i.e. a change of a base by another
one) have the same probability. This probability is the same for all
sites along the DNA sequence. This last assumption can be relaxed by
assuming that the substition rate varies among site following a
gamma distribution which parameter must be given by the user. By
default, no gamma correction is applied. Another assumption is that
the base frequencies are balanced and thus equal to 0.25.
</p>
</li>
<li><p><code>K80</code>: The distance derived by Kimura (1980), sometimes referred
to as &ldquo;Kimura's 2-parameters distance&rdquo;, has the same underlying
assumptions than the Jukes&ndash;Cantor distance except that two kinds of
substitutions are considered: transitions (A &lt;-&gt; G, C &lt;-&gt; T), and
transversions (A &lt;-&gt; C, A &lt;-&gt; T, C &lt;-&gt; G, G &lt;-&gt; T). They are assumed
to have different probabilities. A transition is the substitution of
a purine (C, T) by another one, or the substitution of a pyrimidine
(A, G) by another one. A transversion is the substitution of a
purine by a pyrimidine, or vice-versa. Both transition and
transversion rates are the same for all sites along the DNA
sequence. Jin and Nei (1990) modified the Kimura model to allow for
variation among sites following a gamma distribution. Like for the
Jukes&ndash;Cantor model, the gamma parameter must be given by the
user. By default, no gamma correction is applied.
</p>
</li>
<li><p><code>F81</code>: Felsenstein (1981) generalized the Jukes&ndash;Cantor model
by relaxing the assumption of equal base frequencies. The formulae
used in this function were taken from McGuire et al. (1999).
</p>
</li>
<li><p><code>K81</code>: Kimura (1981) generalized his model (Kimura 1980) by
assuming different rates for two kinds of transversions: A &lt;-&gt; C and
G &lt;-&gt; T on one side, and A &lt;-&gt; T and C &lt;-&gt; G on the other. This is
what Kimura called his &ldquo;three substitution types model&rdquo; (3ST), and
is sometimes referred to as &ldquo;Kimura's 3-parameters distance&rdquo;.
</p>
</li>
<li><p><code>F84</code>: This model generalizes K80 by relaxing the assumption
of equal base frequencies. It was first introduced by Felsenstein in
1984 in Phylip, and is fully described by Felsenstein and Churchill
(1996). The formulae used in this function were taken from McGuire
et al. (1999).
</p>
</li>
<li><p><code>BH87</code>: Barry and Hartigan (1987) developed a distance based
on the observed proportions of changes among the four bases. This
distance is not symmetric.
</p>
</li>
<li><p><code>T92</code>: Tamura (1992) generalized the Kimura model by relaxing
the assumption of equal base frequencies. This is done by taking
into account the bias in G+C content in the sequences. The
substitution rates are assumed to be the same for all sites along
the DNA sequence.
</p>
</li>
<li><p><code>TN93</code>: Tamura and Nei (1993) developed a model which assumes
distinct rates for both kinds of transition (A &lt;-&gt; G versus C &lt;-&gt;
T), and transversions. The base frequencies are not assumed to be
equal and are estimated from the data. A gamma correction of the
inter-site variation in substitution rates is possible.
</p>
</li>
<li><p><code>GG95</code>: Galtier and Gouy (1995) introduced a model where the
G+C content may change through time. Different rates are assumed for
transitons and transversions.
</p>
</li>
<li><p><code>logdet</code>: The Log-Det distance, developed by Lockhart et
al. (1994), is related to BH87. However, this distance is
symmetric. Formulae from Gu and Li (1996) are used.
<code>dist.logdet</code> in <span class="pkg">phangorn</span> uses a different
implementation that gives substantially different distances for
low-diverging sequences.
</p>
</li>
<li><p><code>paralin</code>: Lake (1994) developed the paralinear distance which
can be viewed as another variant of the Barry&ndash;Hartigan distance.
</p>
</li>
<li><p><code>indel</code>: this counts the number of sites where there is an
insertion/deletion gap in one sequence and not in the other.
</p>
</li>
<li><p><code>indelblock</code>: same than before but contiguous gaps are
counted as a single unit. Note that the distance between <code>-A-</code> and
<code>A--</code> is 3 because there are three different blocks of gaps, whereas
the &ldquo;indel&rdquo; distance will be 2.
</p>
</li></ul>


<h3>Value</h3>

<p>an object of class <a href="../../stats/html/dist.html">dist</a> (by default), or a numeric
matrix if <code>as.matrix = TRUE</code>. If <code>model = "BH87"</code>, a numeric
matrix is returned because the Barry&ndash;Hartigan distance is not
symmetric.
</p>
<p>If <code>variance = TRUE</code> an attribute called <code>"variance"</code> is
given to the returned object.
</p>


<h3>Note</h3>

<p>If the sequences are very different, most evolutionary distances are
undefined and a non-finite value (Inf or NaN) is returned. You may do
<code>dist.dna(, model = "raw")</code> to check whether some values are
higher than 0.75.
</p>


<h3>Author(s)</h3>

<p>Emmanuel Paradis</p>


<h3>References</h3>

<p>Barry, D. and Hartigan, J. A. (1987) Asynchronous distance between
homologous DNA sequences. <em>Biometrics</em>, <b>43</b>, 261&ndash;276.
</p>
<p>Felsenstein, J. (1981) Evolutionary trees from DNA sequences: a
maximum likelihood approach. <em>Journal of Molecular Evolution</em>,
<b>17</b>, 368&ndash;376.
</p>
<p>Felsenstein, J. and Churchill, G. A. (1996) A Hidden Markov model
approach to variation among sites in rate of evolution.
<em>Molecular Biology and Evolution</em>, <b>13</b>, 93&ndash;104.
</p>
<p>Galtier, N. and Gouy, M. (1995) Inferring phylogenies from DNA
sequences of unequal base compositions. <em>Proceedings of the
National Academy of Sciences USA</em>, <b>92</b>, 11317&ndash;11321.
</p>
<p>Gu, X. and Li, W.-H. (1996) Bias-corrected paralinear and LogDet
distances and tests of molecular clocks and phylogenies under
nonstationary nucleotide frequencies. <em>Molecular Biology and
Evolution</em>, <b>13</b>, 1375&ndash;1383.
</p>
<p>Jukes, T. H. and Cantor, C. R. (1969) Evolution of protein
molecules. in <em>Mammalian Protein Metabolism</em>, ed. Munro, H. N.,
pp. 21&ndash;132, New York: Academic Press.
</p>
<p>Kimura, M. (1980) A simple method for estimating evolutionary rates of
base substitutions through comparative studies of nucleotide
sequences. <em>Journal of Molecular Evolution</em>, <b>16</b>, 111&ndash;120.
</p>
<p>Kimura, M. (1981) Estimation of evolutionary distances between
homologous nucleotide sequences. <em>Proceedings of the National
Academy of Sciences USA</em>, <b>78</b>, 454&ndash;458.
</p>
<p>Jin, L. and Nei, M. (1990) Limitations of the evolutionary parsimony
method of phylogenetic analysis. <em>Molecular Biology and
Evolution</em>, <b>7</b>, 82&ndash;102.
</p>
<p>Lake, J. A. (1994) Reconstructing evolutionary trees from DNA and
protein sequences: paralinear distances. <em>Proceedings of the
National Academy of Sciences USA</em>, <b>91</b>, 1455&ndash;1459.
</p>
<p>Lockhart, P. J., Steel, M. A., Hendy, M. D. and Penny, D. (1994)
Recovering evolutionary trees under a more realistic model of sequence
evolution. <em>Molecular Biology and Evolution</em>, <b>11</b>,
605&ndash;602.
</p>
<p>McGuire, G., Prentice, M. J. and Wright, F. (1999). Improved error
bounds for genetic distances from DNA sequences. <em>Biometrics</em>,
<b>55</b>, 1064&ndash;1070.
</p>
<p>Tamura, K. (1992) Estimation of the number of nucleotide substitutions
when there are strong transition-transversion and G + C-content
biases. <em>Molecular Biology and Evolution</em>, <b>9</b>, 678&ndash;687.
</p>
<p>Tamura, K. and Nei, M. (1993) Estimation of the number of nucleotide
substitutions in the control region of mitochondrial DNA in humans and
chimpanzees. <em>Molecular Biology and Evolution</em>, <b>10</b>, 512&ndash;526.
</p>


<h3>See Also</h3>

<p><code><a href="read.GenBank.html">read.GenBank</a></code>, <code><a href="read.dna.html">read.dna</a></code>,
<code><a href="write.dna.html">write.dna</a></code>,  <code><a href="DNAbin.html">DNAbin</a></code>,
<code><a href="dist.gene.html">dist.gene</a></code>, <code><a href="cophenetic.phylo.html">cophenetic.phylo</a></code>,
<code><a href="../../stats/html/dist.html">dist</a></code>
</p>

<hr /><div style="text-align: center;">[Package <em>ape</em> version 5.2 <a href="00Index.html">Index</a>]</div>
</body></html>
