<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Translation from DNA to Amino Acid Sequences</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for trans {ape}"><tr><td>trans {ape}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Translation from DNA to Amino Acid Sequences</h2>

<h3>Description</h3>

<p><code>trans</code> translates DNA sequences into amino acids.
<code>complement</code> returns the (reverse) complement sequences.
</p>


<h3>Usage</h3>

<pre>
trans(x, code = 1, codonstart = 1)
complement(x)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>an object of class <code>"DNAbin"</code> (vector, matrix or list).</p>
</td></tr>
<tr valign="top"><td><code>code</code></td>
<td>
<p>an integer value giving the genetic code to be
used. Currently only 1 (standard code) and 2 (vertebrate mitochondrial
code) are supported.</p>
</td></tr>
<tr valign="top"><td><code>codonstart</code></td>
<td>
<p>an integer giving where to start the translation. This
should be 1, 2, or 3, but larger values are accepted and have for
effect to start the translation further within the sequence.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>With <code>trans</code>, if the sequence length is not a multiple of three,
a warning message is printed. Alignment gaps are simply ignored (i.e.,
<code>AG-</code> returns <code>X</code> with no special warning or message). Base
ambiguities are taken into account where relevant: for instance,
<code>GGN</code>, <code>GGA</code>, <code>GGR</code>, etc, all return <code>G</code>.
</p>


<h3>Value</h3>

<p>an object of class <code>"AAbin"</code> or <code>"DNAbin"</code>, respectively.
</p>


<h3>Note</h3>

<p>These functions are equivalent to <code>translate</code> and <code>comp</code> in
the package <span class="pkg">seqinr</span> with the difference that there is no need to
convert the sequences into character strings.
</p>


<h3>Author(s)</h3>

<p>Emmanuel Paradis</p>


<h3>References</h3>

<p><a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes">http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes</a>
</p>


<h3>See Also</h3>

<p><code><a href="AAbin.html">AAbin</a></code>, <code><a href="checkAlignment.html">checkAlignment</a></code>, <code><a href="alview.html">alview</a></code>
</p>


<h3>Examples</h3>

<pre>
data(woodmouse)
X &lt;- trans(woodmouse) # not correct
X2 &lt;- trans(woodmouse, 2) # using the correct code
identical(X, X2)
alview(X[1:2, 1:60]) # some 'Stop' codons (*)
alview(X2[, 1:60])
X2
</pre>

<hr /><div style="text-align: center;">[Package <em>ape</em> version 5.2 <a href="00Index.html">Index</a>]</div>
</body></html>
