<!doctype html>
<html lang="en" class="no-js">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">

	<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,300,600,700' rel='stylesheet' type='text/css'>
	<LINK href="https://www.evinet.org/js/jquery-ui-themes-1.11.2/themes/redmond/jquery-ui-1.11.icon-font.min.css" type="text/css" rel="stylesheet">

	<link rel="stylesheet" href="css/reset.css"> <!-- CSS reset -->
	<link rel="stylesheet" href="css/test_v2.css"> <!-- Resource style -->
	<!--LINK href="../../js/jquery-ui-themes-1.11.2/themes/redmond/jquery-ui.min.css" type="text/css" rel="stylesheet">
	<script src="../../js/jquery-2.1.1.min.js" type="text/javascript"></script>
	<script src="../../js/jquery-ui-1.11.2/jquery-ui.min.js" type="text/javascript"></script-->
	<script src="js/modernizr.js"></script> <!-- Modernizr -->

<title>EviCor FAQ </title>
<style>
.restore_defaults {
	all: initial;
	display: block;
	font-family: "Open Sans", sans-serif;
	margin: 10px;
	
  }
  	TD{font-family: Arial; font-size: 9pt}

	table {
    	font-family: arial, sans-serif;
    	border-collapse: collapse;
    	width: 100%;
	}

	td, th {
    	border: 1px solid black;
    	text-align: center;
    	padding: 8px;
	}

	tr:hover{
   	 background-color: #dddddd;
	}

</style>
</head>

<body>
<header>
	<h1>EviCor: frequently asked questions</h1>
</header>

<!--ul>
<li>Versions</li><li>
Datasets</li><li>
Platforms</li><li>
Metrics</li><li>
Significance interpretation</li><li>
Added value of NEA</li>
</ul-->
<section class="cd-faq">
	<ul class="cd-faq-categories">
		<li><a class="selected" href="#basics">Basics</a></li>
		<li><a href="#data">Data</a></li>
		<li><a href="#usage">Usage</a></li>
		<li><a href="#security">Security and data usage</a></li>
		<li><a href="#mail">Enquires</a></li>
	</ul> <!-- cd-faq-categories -->

	<div class="cd-faq-items">
		<ul id="basics" class="cd-faq-group">
			<li class="cd-faq-title"><h2>Basics</h2></li>
			<li>
				<a class="cd-faq-trigger" href="#0">What EviCor can do for me?</a>
				<div class="cd-faq-content">
					<p>
						Server EviCor was created to provide the biomedical research community with a handy tool for quick look-up at data from large public repositories, The Cancer Genome Atlas (TCGA) and The Cancer Cell Line Encyclopedia (CCLE), and accompanying data on the same samples from other sources, as well as mining basic functional relations via statistical and network tools. None of alternative online resources, to the best of our knowledge, combines quick data access, advanced visualization, and results of correlation and network analysis of the primary data, especially in what concerns response to anti-cancer drugs. 
						In particular, project-specific web sites do not provide an easy way to cross-check findings in alternative in vitro screens and cohorts. EviCor thus represents a first-line tool to test suggested molecular “biomarkers” via biological and statistical validation. It also enables access to pathway-level sample profiles via network enrichment analysis and visualizes it via sub-network views of the pathway correlates. EviCor provides both pre-calculated lists of correlates and direct access to the datasets.
					</p>
				</div> <!-- cd-faq-content -->
			</li>

			<li>
				<a class="cd-faq-trigger" href="#1">What would be an advantage of considering interaction network in my research?</a>
				<div class="cd-faq-content">
					<p>The data integration using global gene networks has the following strengths: 
						<ul>
							<li>The global network includes all major types of molecular interactions, hence the analysis readily incorporates data from different platforms.</li> 
							<li>The problem of cancer heterogeneity at the levels of individual promoters, genes, and proteins is solved by summarizing them to pathways via e.g. the Network Enrichment Analysis, and </li>
							<li>The pathway view enables efficient, low-dimensional statistical analysis and is transparent for biological interpretation. </li>
						</ul>
					</p> 
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#2">What is added value from using network enrichment scores as correlates?</a>
				<div class="cd-faq-content">
					<p>Our work demonstrated advantage of network enrichment analysis (NEA) scores over single gene/protein profiles. The same was true in a comparison against enrichment scores produced by other methods. This was proved in terms of robustness and reproducibility in independent datasets and cohorts:</p>
					<p style="font-size: x-small;">
						Marcela Franco, Ashwini Jeggari, Sylvain Peuget, Franziska B&ouml;ttger, Galina Selivanova, Andrey Alexeyenko
						<span class="highlighted" style="color: #c00;">Prediction of response to anti-cancer drugs becomes robust via network integration of molecular data</span>
						<b><i>Sci Rep</i></b> 9, 2379 (2019) <a class="clickable" href="https://doi.org/10.1038/s41598-019-39019-2">doi: 10.1186/1471-2105-13-226</a>.
					</p>
  					<p>We provide correlates based on the NEA scores in parallel with original gene/protein level data from CCLE and TCGA. Compared to the latter, NEA-based datapoins at the plots are clickable which provides a network context view on the gene sets.</p>
				</div> <!-- cd-faq-content -->
			</li>

		</ul> <!-- cd-faq-group -->

		<ul id="data" class="cd-faq-group">
			<li class="cd-faq-title"><h2>Major public data sources presented at EviCor</h2></li>
			<li>
				<a class="cd-faq-trigger" href="#3">The Cancer Genome Atlas</a>
				<div class="cd-faq-content">
					<p>Ten biggest TCGA cohorts with all their major omics datasets over most of the patient line samples are accompanied with patients' clinical records. In addition, TCGA-based PanCancer projects provided extra data from e.g. leucocyte counts from <i>in silico</i> image analysis over the same samples (Saltz et al., 2018).
					<br><a class="clickable" href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/using-tcga/technology">Omics platforms used by TCGA</a>	
					<br><a class="clickable" href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/studied-cancers">No. of samples in TCGA cohort, per omics platform</a>
						<img src="../../pics/TCGA_10.png" alt="TCGA cohort table" width="96%"> <br> 
					</p>
				</div> <!-- cd-faq-content -->
			</li>						
			<li>
				<a class="cd-faq-trigger" href="#4">The Cancer Cell Line Encyclopedia</a>
				<div class="cd-faq-content">
					<p>The <a class="clickable" href="https://portals.broadinstitute.org/ccle/">CCLE</a> collection of over 1000 cancer cell lines contains both omics (gene expression, gene copy number, point mutation and proteomics) data and results of three  large-scale <i>in vitro</i> drug screens, which inluded more than 600 approved and experimental drugs in total:</p>
					<br>
					<h3>Molecular profiles:</h3>
					<table>				
						<tr><th>Type</th><th>Platform</th><th>No. of genes</th><th>No. of cell lines</th><th>Source</th></tr>  
						<tr><td>GE</td><td>RNAseq</td><td>17893</td><td>1201</td><td><a class="clickable" target="_blank"  href="https://depmap.org/portal/ccle/">CCLE</a></td></tr>  
						<tr><td>GE</td><td>Affymetrix1</td><td>18900</td><td>1034</td><td>CCLE <a class="clickable" target="_blank"  href="https://doi.org/10.1038/nature11003">(Barretina et al., 2012)</a></td></tr>       
						<tr><td>GE</td><td>Affymetrix2</td><td>13981</td><td>622</td>
							<td><a class="clickable" target="_blank"  href="https://www.cancerrxgene.org/">GDSC</a></td></tr>        
						<tr><td>MUT</td><td>CCLE_EXOME_2019</td><td>19525</td><td>1631</td>
							<td><a class="clickable" target="_blank"  href="https://depmap.org/portal/ccle/">CCLE</a></td></tr>           
						<tr><td>MUT</td><td>CCLE_TARGET_2012</td><td>1667</td><td>904</td>
							<td>CCLE <a class="clickable" target="_blank"  href="https://doi.org/10.1038/nature11003">(Barretina et al., 2012)</a></td></tr>             
						<tr><td>MUT</td><td>COSMIC_EXOME_2014</td><td>19759</td><td>1023</td>
							<td><a class="clickable" target="_blank"  href="">COSMIC</a></td></tr>             
						<tr><td>COPY</td><td>SNP6</td><td>21217</td><td>994</td>
							<td>CCLE <a class="clickable" target="_blank"  href="https://doi.org/10.1038/nature11003">(Barretina et al., 2012)</a></td></tr>  
						<tr><td>COPY</td><td>GDSC</td><td>25638</td><td>987</td><td>GDSC <a class="clickable" target="_blank"  href="https://dx.doi.org/10.1016%2Fj.cell.2016.06.017">(Iorio et al., 2016)</a></td></tr>  
						<tr><td>COPY</td><td>CCLE</td><td>23299</td><td>1627</td><td><a class="clickable" target="_blank"  href="https://depmap.org/portal/ccle/">CCLE</a></td></tr> 
						<tr><td>METH</td><td>GSE68379</td><td>485512</td><td>1028</td><td>GDSC <a class="clickable" target="_blank"  href="https://dx.doi.org/10.1016%2Fj.cell.2016.06.017">(Iorio et al., 2016)</a></td></tr> 				
					</table>
					<br>

					<h3>Drug screens:</h3>
<table>
<tr><th>Screen</th><th>No. of drugs</th><th>No. of cell lines</th><th>Source</th></tr>
<tr><td>GDSC1</td><td>304</td><td>988</td><td><a class="clickable" target="_blank"  href="https://www.cancerrxgene.org/">GDSC</a></td></tr>
<tr><td>GDSC2</td><td>169</td><td>810</td><td><a class="clickable" target="_blank"  href="https://www.cancerrxgene.org/">GDSC</a></td></tr>
<tr><td>CTRPv2.0</td><td>545</td><td>887</td><td><a class="clickable" target="_blank"  href="https://depmap.org/portal/ccle/">CTRPv2.0</a></td></tr>	
						<!--tr><td>ACT</td><td>25</td><td>30</td><td><a class="clickable" target="_blank"  href="https://depmap.org/portal/ccle/">CCLE</a></td></tr> 	
						<tr><td>Basu2013</td><td>203</td><td>242</td><td><a class="clickable" target="_blank"  href="http://dx.doi.org/10.1016/j.cell.2013.08.003">CTD2</a></td></tr> 	
						<tr><td>Garnett2012</td><td>138</td><td>714</td><td><a class="clickable" target="_blank"  href="https://www.cancerrxgene.org/">GDSC</a></td></tr> 	
						<tr><td>Barretina2012</td><td>24</td><td>504</td><td><a class="clickable" target="_blank"  href="https://doi.org/10.1038/nature11003">CCLE</a></td></tr--> 	 
</table>
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#5">Which IDs can be used for querying?</a>
				<div class="cd-faq-content">
					<p> 
						The tables of molecular variables usually store gene symbols, a.k.a. HUGO gene symbols, which mostly coincide with protein HUPO gene symbols. Alternatively, you can query by SwissProt ID and AC, or ENSEMBL gene (ENSG*), protein (ENSP*), and transcript (ENST*) IDs. Proteomics RPPA arrays can be queried by both original antibody IDs and respective HUGO symbols. Drug names are standard ids presented by screen data providers and TCGA (should be avaialable with and without brackets and dashes in the middle). Autocomplete should quickly prompt if the ID exists, usually after typing in 3-4 first characters. In tab "Correlates of drug response", an ID would be shown only if its respective correlations were save as significant in the database. 
					</p>
				</div> <!-- cd-faq-content -->
			</li>
		</ul> <!-- cd-faq-group -->

		<ul id="usage" class="cd-faq-group">
			<li class="cd-faq-title"><h2>Using the web server</h2></li>
			<li>
				<a class="cd-faq-trigger" href="#5">Finding pre-calculated correlates</a>
				<div class="cd-faq-content">
					<p> 
						The table in second tab allows extracting lists of molecular correlates from particular omics platforms for specific drugs. In CCLE, we calculated correlates of drug sensitivity <i>in vitro</i> in the commonly accepted 2-factorial linear model, where the tissue of origin served a covariate. Therefore, the other model term, called "feature" represents a "net" effect of the molecular feature. <br>
						In TCGA, the 2-factorial models with interaction terms allowed comparison of disease outcome given high vs. low molecular feature score between groups of patients who differed in having/not having the drug treatment in the clinical history. Thus the significant (i.e. low p-value) "interaction" term in the model points to a predictive potential of the molecular feature with regard of the treatment outcome.
						Demos on the top of the main page can help to start.
					</p>
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#6">Exploring relations between different variables</a>
				<div class="cd-faq-content">
					<p> 
						Tab 2 provides access to most of the datasets in TCGA  and CCLE. One, two, or three molecular and/or phenotypic variables can be selected and represented via different plot types. 
						By clicking '+' in the end of each dimension line, one can add up dimensions to the plot (the last, 3rd one would then be represented via colors or shapes). If a certain data-plot combination is not enabled, it is recommended to remove the lines and start again in a different order.
						Demos on the top of the main page can help to start.
					</p>
				</div> <!-- cd-faq-content -->
			</li>				
			<li>
				<a class="cd-faq-trigger" href="#6">On what scale is the data?</a>
				<div class="cd-faq-content">
					<p> 
						The scales called "original" are, in fact, values after a certain transformation. <br>
						1) RNA-seq (FPKM), microarray, and proteomics values are log-transformed either by us or already in the original source.<br>
						2) Gene copy number values are log-transformed so that "0" corresponds to the normal number of copies.<br>
						3) Mutations are presented in the binary format, i.e. any line reported in the original MAF files is considered a gene mutation, regardless of its intron/exon position or functional impact.<br>
						4) Methylation in is presented as beta-values, i.e. percent of methylated CG sites in the gene promoter region. The optional M-values are logit units log(beta/(a-beta)). These are normally distributed and more suitable for the of most statistical models.<br>
						5) Drug sensitivity screens are presented as AUC (are under S-shaped curve of cell viability at a range of dilutions), transformed to close-to-normal distribution by inverse quantile normalizationa across cell lines.<br>
						Most of the other variable are given as there were originally provided.<br>
						Variable representation in the plots can be changed via transformation to log or square root scales. 
					</p>
				</div> <!-- cd-faq-content -->
			</li>	
			<li>
				<a class="cd-faq-trigger" href="#7">Viewing and saving plots</a>
				<div class="cd-faq-content">
					<p>
						Most plots on the site are created with <a href=https://plot.ly/>plotly</a> library and therefore highly interactive. If a scatter plot has several datalines, such as colored categories, then you can turn some of them of (same goes for pie charts). If a scatter plots presents network enrichment scores, clicking on each point will display a sub-network of connections between the respective pathway and sample-specific genes.
						<br>
						You can save plots either as PNG files (via button in the upper-right window corner) or as a URL (via "Get link" button under the plot). URLs can be used for retrieving plots in separate browser windows or embedded elsewhere.
					</p>
				</div>
			</li>
			<li>
				<a class="cd-faq-trigger" href="#8">Building multivariate models</a>
				<div class="cd-faq-content">
					<p>
						Tab 3 of this site would allow you to build multivariate models. For this, choose what you want to predict (s.c. dependent variable). For TCGA data you can choose which samples will be used for dependent and independent variables (e.g. only samples from primary solid tumor, see code explanation <a href=https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes target="_blank">here</a>), for CCLE you can choose cell lines by choosing their tissue of origin. If you will choose just one variable - you will get a 2D plot instead of multidimensional model.
						You can choose if you want model verification or not. After your model is built, you can explore it, download it (using the respective button), view coefficients. Table with coefficients contains their weights and "Plot" button for each of them. Click this button if you want to plot data behind this coefficient (it will use response variable data for another axis).
					</p>
				</div>
			</li>
		</ul> <!-- cd-faq-group -->
		
		<ul id="security" class="cd-faq-group">
			<li class="cd-faq-title"><h2>Security and data usage</h2></li>
			<li>
				<a class="cd-faq-trigger" href="#9">Is my connection secure?</a>
				<div class="cd-faq-content">
					<p>Yes. This server uses https connection.</p>
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#10">Why do I have to accept cookies?</a>
				<div class="cd-faq-content">
					<p>Cookies are used for memorizing your latest settings. Due to this feature you would be able to reload the site and some settings, such as chosen cohort, datatypes, platforms etc., will be restored automatically.</p>
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#11">How the cookies would be used?</a>
				<div class="cd-faq-content">
					<p>EviCor uses cookies only for improving user experience with website. We don't have access to your personal data and don't store any information about user's actions, except some error messages. You have to give your permission for us to in order to enable using cookies. This site also uses Google Analytics and all related cookies are handled by Google according to their policy.</p>
				</div> <!-- cd-faq-content -->
			</li>
			<li>
				<a class="cd-faq-trigger" href="#12">Do you collect user data?</a>
				<div class="cd-faq-content">
					<p>No, as said: we don't collect user data. Some rare situations (such as severe errors) trigger report generations. These reports include the following: date and time of error registration, error description, options which led to error (values of selectors etc.), and information about the user's browser (version etc.). EviCor does not store IP addressed or any similiar information.</p>
				</div> <!-- cd-faq-content -->
			</li>
		</ul> <!-- cd-faq-group -->
	
		<ul id="mail" class="cd-faq-group">
			<li class="cd-faq-title"><h2>Enquires</h2></li>
			<li>
				<a class="cd-faq-trigger" href="#13">Found problems?</a>
				<div class="cd-faq-content">
					<a href="mailto:andrej.alekseenko@scilifelab.se">Send us email</a>
				</div> <!-- cd-faq-content -->
			</li>
		</ul> <!-- cd-faq-group -->
	</div> <!-- cd-faq-items -->
	<a href="#0" class="cd-close-panel">Close</a>
</section> <!-- cd-faq -->

<script src="js/jquery-2.1.1.js"></script>
<script src="js/jquery.mobile.custom.min.js"></script>
<script src="js/main.js"></script> <!-- Resource jQuery -->
</body>
</html>

