<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Utility functions related to sequence alignment</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for align-utils {Biostrings}"><tr><td>align-utils {Biostrings}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Utility functions related to sequence alignment</h2>

<h3>Description</h3>

<p>A variety of different functions used to deal with sequence alignments.
</p>


<h3>Usage</h3>

<pre>
nedit(x) # also nmatch and nmismatch

mismatchTable(x, shiftLeft=0L, shiftRight=0L, ...)
mismatchSummary(x, ...)
## S4 method for signature 'AlignedXStringSet0'
coverage(x, shift=0L, width=NULL, weight=1L)
## S4 method for signature 'PairwiseAlignmentsSingleSubject'
coverage(x, shift=0L, width=NULL, weight=1L)
compareStrings(pattern, subject)

## S4 method for signature 'PairwiseAlignmentsSingleSubject'
consensusMatrix(x,
                as.prob=FALSE, shift=0L, width=NULL,
                baseOnly=FALSE, gapCode="-", endgapCode="-")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <code>character</code> vector or matrix, <code>XStringSet</code>, <code>XStringViews</code>,
<code>PairwiseAlignments</code>, or <code>list</code> of FASTA records containing
the equal-length strings.
</p>
</td></tr>
<tr valign="top"><td><code>shiftLeft, shiftRight</code></td>
<td>

<p>Non-positive and non-negative integers respectively that specify how many
preceding and succeeding characters to and from the mismatch position to
include in the mismatch substrings.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Further arguments to be passed to or from other methods.
</p>
</td></tr>
<tr valign="top"><td><code>shift, width</code></td>
<td>

<p>See <code>?<a href="../../IRanges/html/coverage.html">coverage</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>weight</code></td>
<td>

<p>An integer vector specifying how much each element in <code>x</code> counts.
</p>
</td></tr>
<tr valign="top"><td><code>pattern, subject</code></td>
<td>

<p>The strings to compare. Can be of type <code>character</code>, <code>XString</code>,
<code>XStringSet</code>, <code>AlignedXStringSet</code>, or, in the case of
<code>pattern</code>, <code>PairwiseAlignments</code>. If <code>pattern</code> is a
<code>PairwiseAlignments</code> object, then <code>subject</code> must be
missing.
</p>
</td></tr>
<tr valign="top"><td><code>as.prob</code></td>
<td>

<p>If <code>TRUE</code> then probabilities are reported,
otherwise counts (the default).
</p>
</td></tr>
<tr valign="top"><td><code>baseOnly</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>.
If <code>TRUE</code>, the returned vector only contains frequencies for the
letters in the &quot;base&quot; alphabet i.e. &quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot; if <code>x</code>
is a &quot;DNA input&quot;, and &quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;U&quot; if <code>x</code> is &quot;RNA input&quot;.
When <code>x</code> is a <a href="XString-class.html">BString</a> object (or an <a href="XStringViews-class.html">XStringViews</a>
object with a <a href="XString-class.html">BString</a> subject, or a <a href="XStringSet-class.html">BStringSet</a> object),
then the <code>baseOnly</code> argument is ignored.
</p>
</td></tr>
<tr valign="top"><td><code>gapCode, endgapCode</code></td>
<td>

<p>The codes in the appropriate <code><a href="XString-class.html">alphabet</a></code> to use for the internal
and end gaps.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>mismatchTable</code>:  a data.frame containing the positions and substrings
of the mismatches for the <code>AlignedXStringSet</code> or
<code>PairwiseAlignments</code> object.
</p>
<p><code>mismatchSummary</code>:  a list of data.frame objects containing counts and
frequencies of the mismatches for the <code>AlignedXStringSet</code> or
<code>PairwiseAlignmentsSingleSubject</code> object.
</p>
<p><code>compareStrings</code> combines two equal-length strings that are assumed to be
aligned into a single character string containing that replaces mismatches
with <code>"?"</code>, insertions with <code>"+"</code>, and deletions with <code>"-"</code>.
</p>


<h3>See Also</h3>

<p><code><a href="pairwiseAlignment.html">pairwiseAlignment</a></code>,
<code><a href="letterFrequency.html">consensusMatrix</a></code>,
<a href="XString-class.html">XString-class</a>, <a href="XStringSet-class.html">XStringSet-class</a>, <a href="XStringViews-class.html">XStringViews-class</a>,
<a href="AlignedXStringSet-class.html">AlignedXStringSet-class</a>, <a href="PairwiseAlignments-class.html">PairwiseAlignments-class</a>,
<a href="match-utils.html">match-utils</a>
</p>


<h3>Examples</h3>

<pre>
  ## Compare two globally aligned strings
  string1 &lt;- "ACTTCACCAGCTCCCTGGCGGTAAGTTGATC---AAAGG---AAACGCAAAGTTTTCAAG"
  string2 &lt;- "GTTTCACTACTTCCTTTCGGGTAAGTAAATATATAAATATATAAAAATATAATTTTCATC"
  compareStrings(string1, string2)

  ## Create a consensus matrix
  nw1 &lt;-
    pairwiseAlignment(AAStringSet(c("HLDNLKGTF", "HVDDMPNAL")), AAString("SMDDTEKMSMKL"),
      substitutionMatrix = "BLOSUM50", gapOpening = 3, gapExtension = 1)
  consensusMatrix(nw1)

  ## Examine the consensus between the bacteriophage phi X174 genomes
  data(phiX174Phage)
  phageConsmat &lt;- consensusMatrix(phiX174Phage, baseOnly = TRUE)
  phageDiffs &lt;- which(apply(phageConsmat, 2, max) &lt; length(phiX174Phage))
  phageDiffs
  phageConsmat[,phageDiffs]
</pre>

<hr /><div style="text-align: center;">[Package <em>Biostrings</em> version 2.50.1 <a href="00Index.html">Index</a>]</div>
</body></html>
