<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Matching a dictionary of patterns against a reference</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for matchPDict {Biostrings}"><tr><td>matchPDict {Biostrings}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Matching a dictionary of patterns against a reference</h2>

<h3>Description</h3>

<p>A set of functions for finding all the occurrences (aka &quot;matches&quot;
or &quot;hits&quot;) of a set of patterns (aka the dictionary) in a reference
sequence or set of reference sequences (aka the subject)
</p>
<p>The following functions differ in what they return: <code>matchPDict</code>
returns the &quot;where&quot; information i.e. the positions in the subject of all the
occurrences of every pattern; <code>countPDict</code> returns the &quot;how many
times&quot; information i.e. the number of occurrences for each pattern;
and <code>whichPDict</code> returns the &quot;who&quot; information i.e. which patterns
in the input dictionary have at least one match.
</p>
<p><code>vcountPDict</code> and <code>vwhichPDict</code> are vectorized versions
of <code>countPDict</code> and <code>whichPDict</code>, respectively, that is,
they work on a set of reference sequences in a vectorized fashion.
</p>
<p>This man page shows how to use these functions (aka the <code>*PDict</code>
functions) for exact matching of a constant width dictionary i.e.
a dictionary where all the patterns have the same length (same number
of nucleotides).
</p>
<p>See <code>?`<a href="matchPDict-inexact.html">matchPDict-inexact</a>`</code> for how to use these functions
for inexact matching or when the original dictionary has a variable width.
</p>


<h3>Usage</h3>

<pre>
matchPDict(pdict, subject,
           max.mismatch=0, min.mismatch=0, with.indels=FALSE, fixed=TRUE,
           algorithm="auto", verbose=FALSE)
countPDict(pdict, subject,
           max.mismatch=0, min.mismatch=0, with.indels=FALSE, fixed=TRUE,
           algorithm="auto", verbose=FALSE)
whichPDict(pdict, subject,
           max.mismatch=0, min.mismatch=0, with.indels=FALSE, fixed=TRUE,
           algorithm="auto", verbose=FALSE)

vcountPDict(pdict, subject,
            max.mismatch=0, min.mismatch=0, with.indels=FALSE, fixed=TRUE,
            algorithm="auto", collapse=FALSE, weight=1L,
            verbose=FALSE, ...)
vwhichPDict(pdict, subject,
            max.mismatch=0, min.mismatch=0, with.indels=FALSE, fixed=TRUE,
            algorithm="auto", verbose=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>pdict</code></td>
<td>

<p>A <a href="PDict-class.html">PDict</a> object containing the preprocessed dictionary.
</p>
<p>All these functions also work with a dictionary that has not been
preprocessed (in other words, the <code>pdict</code> argument can receive
an <a href="XStringSet-class.html">XStringSet</a> object). Of course, it won't be as fast as with
a preprocessed dictionary, but it will generally be slightly faster
than using <code><a href="matchPattern.html">matchPattern</a></code>/<code><a href="matchPattern.html">countPattern</a></code>
or <code><a href="matchPattern.html">vmatchPattern</a></code>/<code><a href="matchPattern.html">vcountPattern</a></code> in a
&quot;lapply/sapply loop&quot;, because, here, looping is done at the C-level.
However, by using a non-preprocessed dictionary, many of the
restrictions that apply to preprocessed dictionaries don't apply
anymore. For example, the dictionary doesn't need to be rectangular
or to be a <a href="XStringSet-class.html">DNAStringSet</a> object: it can be any type of
<a href="XStringSet-class.html">XStringSet</a> object and have a variable width.
</p>
</td></tr>
<tr valign="top"><td><code>subject</code></td>
<td>

<p>An <a href="XString-class.html">XString</a> or <a href="MaskedXString-class.html">MaskedXString</a> object containing the
subject sequence for <code>matchPDict</code>, <code>countPDict</code> and
<code>whichPDict</code>.
</p>
<p>An <a href="XStringSet-class.html">XStringSet</a> object containing the subject sequences
for <code>vcountPDict</code> and <code>vwhichPDict</code>.
</p>
<p>If <code>pdict</code> is a <a href="PDict-class.html">PDict</a> object (i.e. a preprocessed
dictionary), then <code>subject</code> must be of base class <a href="DNAString-class.html">DNAString</a>.
Otherwise, <code>subject</code> must be of the same base class as <code>pdict</code>.
</p>
</td></tr>
<tr valign="top"><td><code>max.mismatch, min.mismatch</code></td>
<td>

<p>The maximum and minimum number of mismatching letters allowed (see
<code>?<a href="lowlevel-matching.html">isMatchingAt</a></code> for the details).
This man page focuses on exact matching of a constant width
dictionary so <code>max.mismatch=0</code> in the examples below.
See <code>?`<a href="matchPDict-inexact.html">matchPDict-inexact</a>`</code> for inexact matching.
</p>
</td></tr>
<tr valign="top"><td><code>with.indels</code></td>
<td>

<p>Only supported by <code>countPDict</code>, <code>whichPDict</code>,
<code>vcountPDict</code> and <code>vwhichPDict</code> at the moment, and only
when the input dictionary is non-preprocessed (i.e. XStringSet).
</p>
<p>If <code>TRUE</code> then indels are allowed. In that case, <code>min.mismatch</code>
must be <code>0</code> and <code>max.mismatch</code> is interpreted as the maximum
&quot;edit distance&quot; allowed between any pattern and any of its matches.
See <code>?`<a href="matchPattern.html">matchPattern</a>`</code> for more information.
</p>
</td></tr>
<tr valign="top"><td><code>fixed</code></td>
<td>

<p>Whether IUPAC ambiguity codes should be interpreted literally or not
(see <code>?<a href="lowlevel-matching.html">isMatchingAt</a></code> for more information).
This man page focuses on exact matching of a constant width
dictionary so <code>fixed=TRUE</code> in the examples below.
See <code>?`<a href="matchPDict-inexact.html">matchPDict-inexact</a>`</code> for inexact matching.
</p>
</td></tr>
<tr valign="top"><td><code>algorithm</code></td>
<td>

<p>Ignored if <code>pdict</code> is a preprocessed dictionary (i.e.
a <a href="PDict-class.html">PDict</a> object). Otherwise, can be one of the following:
<code>"auto"</code>, <code>"naive-exact"</code>, <code>"naive-inexact"</code>,
<code>"boyer-moore"</code> or <code>"shift-or"</code>.
See <code>?<a href="matchPattern.html">matchPattern</a></code> for more information.
Note that <code>"indels"</code> is not supported for now.
</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>.
</p>
</td></tr>
<tr valign="top"><td><code>collapse, weight</code></td>
<td>

<p><code>collapse</code> must be <code>FALSE</code>, <code>1</code>, or <code>2</code>.
</p>
<p>If <code>collapse=FALSE</code> (the default), then <code>weight</code> is ignored
and <code>vcountPDict</code> returns the full matrix of counts (<code>M0</code>).
If <code>collapse=1</code>, then <code>M0</code> is collapsed &quot;horizontally&quot;
i.e. it is turned into a vector with <code>length</code> equal to
<code>length(pdict)</code>.
If <code>weight=1L</code> (the default), then this vector is defined by
<code>rowSums(M0)</code>.
If <code>collapse=2</code>, then <code>M0</code> is collapsed &quot;vertically&quot;
i.e. it is turned into a vector with <code>length</code> equal to
<code>length(subject)</code>.
If <code>weight=1L</code> (the default), then this vector is defined by
<code>colSums(M0)</code>.
</p>
<p>If <code>collapse=1</code> or <code>collapse=2</code>, then the elements in
<code>subject</code> (<code>collapse=1</code>) or in <code>pdict</code> (<code>collapse=2</code>)
can be weighted thru the <code>weight</code> argument.
In that case, the returned vector is defined by
<code>M0 %*% rep(weight, length.out=length(subject))</code>
and <code>rep(weight, length.out=length(pdict)) %*% M0</code>,
respectively.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments for methods.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>In this man page, we assume that you know how to preprocess a dictionary
of DNA patterns that can then be used with any of the <code>*PDict</code>
functions described here. Please see <code>?<a href="PDict-class.html">PDict</a></code> if you don't.
</p>
<p>When using the <code>*PDict</code> functions for exact matching of a constant
width dictionary, the standard way to preprocess the original dictionary
is by calling the <code><a href="PDict-class.html">PDict</a></code> constructor on it with no extra
arguments. This returns the preprocessed dictionary in a <a href="PDict-class.html">PDict</a>
object that can be used with any of the <code>*PDict</code> functions.
</p>


<h3>Value</h3>

<p>If <code>M</code> denotes the number of patterns in the <code>pdict</code>
argument (<code>M &lt;- length(pdict)</code>), then <code>matchPDict</code> returns
an <a href="MIndex-class.html">MIndex</a> object of length <code>M</code>,
and <code>countPDict</code> an integer vector of length <code>M</code>.
</p>
<p><code>whichPDict</code> returns an integer vector made of the indices of the
patterns in the <code>pdict</code> argument that have at least one match.
</p>
<p>If <code>N</code> denotes the number of sequences in the <code>subject</code>
argument (<code>N &lt;- length(subject)</code>), then <code>vcountPDict</code>
returns an integer matrix with <code>M</code> rows and <code>N</code> columns,
unless the <code>collapse</code> argument is used. In that case, depending
on the type of <code>weight</code>, an integer or numeric vector is returned
(see above for the details).
</p>
<p><code>vwhichPDict</code> returns a list of <code>N</code> integer vectors.
</p>


<h3>Author(s)</h3>

<p>H. Pagès</p>


<h3>References</h3>

<p>Aho, Alfred V.; Margaret J. Corasick (June 1975). &quot;Efficient string
matching: An aid to bibliographic search&quot;.
Communications of the ACM 18 (6): 333-340.
</p>


<h3>See Also</h3>

<p><a href="PDict-class.html">PDict-class</a>,
<a href="MIndex-class.html">MIndex-class</a>,
<a href="matchPDict-inexact.html">matchPDict-inexact</a>,
<code><a href="lowlevel-matching.html">isMatchingAt</a></code>,
<code><a href="match-utils.html">coverage,MIndex-method</a></code>,
<code><a href="matchPattern.html">matchPattern</a></code>,
<code><a href="letterFrequency.html">alphabetFrequency</a></code>,
<a href="XStringSet-class.html">DNAStringSet-class</a>,
<a href="XStringViews-class.html">XStringViews-class</a>,
<a href="MaskedXString-class.html">MaskedDNAString-class</a>
</p>


<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. A SIMPLE EXAMPLE OF EXACT MATCHING
## ---------------------------------------------------------------------

## Creating the pattern dictionary:
library(drosophila2probe)
dict0 &lt;- DNAStringSet(drosophila2probe)
dict0                                # The original dictionary.
length(dict0)                        # Hundreds of thousands of patterns.
pdict0 &lt;- PDict(dict0)               # Store the original dictionary in
                                     # a PDict object (preprocessing).

## Using the pattern dictionary on chromosome 3R:
library(BSgenome.Dmelanogaster.UCSC.dm3)
chr3R &lt;- Dmelanogaster$chr3R         # Load chromosome 3R
chr3R
mi0 &lt;- matchPDict(pdict0, chr3R)     # Search...

## Looking at the matches:
start_index &lt;- startIndex(mi0)       # Get the start index.
length(start_index)                  # Same as the original dictionary.
start_index[[8220]]                  # Starts of the 8220th pattern.
end_index &lt;- endIndex(mi0)           # Get the end index.
end_index[[8220]]                    # Ends of the 8220th pattern.
nmatch_per_pat &lt;- elementNROWS(mi0)  # Get the number of matches per pattern.
nmatch_per_pat[[8220]]
mi0[[8220]]                          # Get the matches for the 8220th pattern.
start(mi0[[8220]])                   # Equivalent to startIndex(mi0)[[8220]].
sum(nmatch_per_pat)                  # Total number of matches.
table(nmatch_per_pat)
i0 &lt;- which(nmatch_per_pat == max(nmatch_per_pat))
pdict0[[i0]]                         # The pattern with most occurrences.
mi0[[i0]]                            # Its matches as an IRanges object.
Views(chr3R, mi0[[i0]])              # And as an XStringViews object.

## Get the coverage of the original subject:
cov3R &lt;- as.integer(coverage(mi0, width=length(chr3R)))
max(cov3R)
mean(cov3R)
sum(cov3R != 0) / length(cov3R)      # Only 2.44% of chr3R is covered.
if (interactive()) {
  plotCoverage &lt;- function(cx, start, end)
  {
    plot.new()
    plot.window(c(start, end), c(0, 20))
    axis(1)
    axis(2)
    axis(4)
    lines(start:end, cx[start:end], type="l")
  }
  plotCoverage(cov3R, 27600000, 27900000)
}

## ---------------------------------------------------------------------
## B. NAMING THE PATTERNS
## ---------------------------------------------------------------------

## The names of the original patterns, if any, are propagated to the
## PDict and MIndex objects:
names(dict0) &lt;- mkAllStrings(letters, 4)[seq_len(length(dict0))]
dict0
dict0[["abcd"]]
pdict0n &lt;- PDict(dict0)
names(pdict0n)[1:30]
pdict0n[["abcd"]]
mi0n &lt;- matchPDict(pdict0n, chr3R)
names(mi0n)[1:30]
mi0n[["abcd"]]

## This is particularly useful when unlisting an MIndex object:
unlist(mi0)[1:10]
unlist(mi0n)[1:10]  # keep track of where the matches are coming from

## ---------------------------------------------------------------------
## C. PERFORMANCE
## ---------------------------------------------------------------------

## If getting the number of matches is what matters only (without
## regarding their positions), then countPDict() will be faster,
## especially when there is a high number of matches:

nmatch_per_pat0 &lt;- countPDict(pdict0, chr3R)
stopifnot(identical(nmatch_per_pat0, nmatch_per_pat))

if (interactive()) {
  ## What's the impact of the dictionary width on performance?
  ## Below is some code that can be used to figure out (will take a long
  ## time to run). For different widths of the original dictionary, we
  ## look at:
  ##   o pptime: preprocessing time (in sec.) i.e. time needed for
  ##             building the PDict object from the truncated input
  ##             sequences;
  ##   o nnodes: nb of nodes in the resulting Aho-Corasick tree;
  ##   o nupatt: nb of unique truncated input sequences;
  ##   o matchtime: time (in sec.) needed to find all the matches;
  ##   o totalcount: total number of matches.
  getPDictStats &lt;- function(dict, subject)
  {
    ans_width &lt;- width(dict[1])
    ans_pptime &lt;- system.time(pdict &lt;- PDict(dict))[["elapsed"]]
    pptb &lt;- pdict@threeparts@pptb
    ans_nnodes &lt;- nnodes(pptb)
    ans_nupatt &lt;- sum(!duplicated(pdict))
    ans_matchtime &lt;- system.time(
                       mi0 &lt;- matchPDict(pdict, subject)
                     )[["elapsed"]]
    ans_totalcount &lt;- sum(elementNROWS(mi0))
    list(
      width=ans_width,
      pptime=ans_pptime,
      nnodes=ans_nnodes,
      nupatt=ans_nupatt,
      matchtime=ans_matchtime,
      totalcount=ans_totalcount
    )
  }
  stats &lt;- lapply(8:25,
               function(width)
                   getPDictStats(DNAStringSet(dict0, end=width), chr3R))
  stats &lt;- data.frame(do.call(rbind, stats))
  stats
}

## ---------------------------------------------------------------------
## D. USING A NON-PREPROCESSED DICTIONARY
## ---------------------------------------------------------------------

dict3 &lt;- DNAStringSet(mkAllStrings(DNA_BASES, 3))  # all trinucleotides
dict3
pdict3 &lt;- PDict(dict3)

## The 3 following calls are equivalent (from faster to slower):
res3a &lt;- countPDict(pdict3, chr3R)
res3b &lt;- countPDict(dict3, chr3R)
res3c &lt;- sapply(dict3,
             function(pattern) countPattern(pattern, chr3R))
stopifnot(identical(res3a, res3b))
stopifnot(identical(res3a, res3c))

## One reason for using a non-preprocessed dictionary is to get rid of
## all the constraints associated with preprocessing, e.g., when
## preprocessing with PDict(), the input dictionary must be DNA and a
## Trusted Band must be defined (explicitly or implicitly).
## See '?PDict' for more information about these constraints.
## In particular, using a non-preprocessed dictionary can be
## useful for the kind of inexact matching that can't be achieved
## with a PDict object (if performance is not an issue).
## See '?`matchPDict-inexact`' for more information about inexact
## matching.

dictD &lt;- xscat(dict3, "N", reverseComplement(dict3))

## The 2 following calls are equivalent (from faster to slower):
resDa &lt;- matchPDict(dictD, chr3R, fixed=FALSE)
resDb &lt;- sapply(dictD,
                function(pattern)
                  matchPattern(pattern, chr3R, fixed=FALSE))
stopifnot(all(sapply(seq_len(length(dictD)),
                     function(i)
                       identical(resDa[[i]], as(resDb[[i]], "IRanges")))))

## A non-preprocessed dictionary can be of any base class i.e. BString,
## RNAString, and AAString, in addition to DNAString:
matchPDict(AAStringSet(c("DARC", "EGH")), AAString("KMFPRNDEGHSTTWTEE"))

## ---------------------------------------------------------------------
## E. vcountPDict()
## ---------------------------------------------------------------------

## Load Fly upstream sequences (i.e. the sequences 2000 bases upstream of
## annotated transcription starts):
dm3_upstream_filepath &lt;- system.file("extdata",
                                     "dm3_upstream2000.fa.gz",
                                     package="Biostrings")
dm3_upstream &lt;- readDNAStringSet(dm3_upstream_filepath)
dm3_upstream

subject &lt;- dm3_upstream[1:100]
mat1 &lt;- vcountPDict(pdict0, subject)
dim(mat1)  # length(pdict0) x length(subject)
nhit_per_probe &lt;- rowSums(mat1)
table(nhit_per_probe)

## Without vcountPDict(), 'mat1' could have been computed with:
mat2 &lt;- sapply(unname(subject), function(x) countPDict(pdict0, x))
stopifnot(identical(mat1, mat2))
## but using vcountPDict() is faster (10x or more, depending of the
## average length of the sequences in 'subject').

if (interactive()) {
  ## This will fail (with message "allocMatrix: too many elements
  ## specified") because, on most platforms, vectors and matrices in R
  ## are limited to 2^31 elements:
  subject &lt;- dm3_upstream
  vcountPDict(pdict0, subject)
  length(pdict0) * length(dm3_upstream)
  1 * length(pdict0) * length(dm3_upstream)  # &gt; 2^31
  ## But this will work:
  nhit_per_seq &lt;- vcountPDict(pdict0, subject, collapse=2)
  sum(nhit_per_seq &gt;= 1)  # nb of subject sequences with at least 1 hit
  table(nhit_per_seq)  # max is 74
  which.max(nhit_per_seq)  # 1133
  sum(countPDict(pdict0, subject[[1133]]))  # 74
}

## ---------------------------------------------------------------------
## F. RELATIONSHIP BETWEEN vcountPDict(), countPDict() AND
## vcountPattern()
## ---------------------------------------------------------------------
subject &lt;- dm3_upstream

## The 4 following calls are equivalent (from faster to slower):
mat3a &lt;- vcountPDict(pdict3, subject)
mat3b &lt;- vcountPDict(dict3, subject)
mat3c &lt;- sapply(dict3,
                function(pattern) vcountPattern(pattern, subject))
mat3d &lt;- sapply(unname(subject),
                function(x) countPDict(pdict3, x))
stopifnot(identical(mat3a, mat3b))
stopifnot(identical(mat3a, t(mat3c)))
stopifnot(identical(mat3a, mat3d))

## The 3 following calls are equivalent (from faster to slower):
nhitpp3a &lt;- vcountPDict(pdict3, subject, collapse=1)  # rowSums(mat3a)
nhitpp3b &lt;- vcountPDict(dict3, subject, collapse=1)
nhitpp3c &lt;- sapply(dict3,
                   function(pattern) sum(vcountPattern(pattern, subject)))
stopifnot(identical(nhitpp3a, nhitpp3b))
stopifnot(identical(nhitpp3a, nhitpp3c))

## The 3 following calls are equivalent (from faster to slower):
nhitps3a &lt;- vcountPDict(pdict3, subject, collapse=2)  # colSums(mat3a)
nhitps3b &lt;- vcountPDict(dict3, subject, collapse=2)
nhitps3c &lt;- sapply(unname(subject),
                   function(x) sum(countPDict(pdict3, x)))
stopifnot(identical(nhitps3a, nhitps3b))
stopifnot(identical(nhitps3a, nhitps3c))

## ---------------------------------------------------------------------
## G. vwhichPDict()
## ---------------------------------------------------------------------
subject &lt;- dm3_upstream

## The 4 following calls are equivalent (from faster to slower):
vwp3a &lt;- vwhichPDict(pdict3, subject)
vwp3b &lt;- vwhichPDict(dict3, subject)
vwp3c &lt;- lapply(seq_len(ncol(mat3a)), function(j) which(mat3a[ , j] != 0L))
vwp3d &lt;- lapply(unname(subject), function(x) whichPDict(pdict3, x))
stopifnot(identical(vwp3a, vwp3b))
stopifnot(identical(vwp3a, vwp3c))
stopifnot(identical(vwp3a, vwp3d))

table(sapply(vwp3a, length))
which.min(sapply(vwp3a, length))
## Get the trinucleotides not represented in upstream sequence 21823:
dict3[-vwp3a[[21823]]]  # 2 trinucleotides

## Sanity check:
tnf &lt;- trinucleotideFrequency(subject[[21823]])
stopifnot(all(names(tnf)[tnf == 0] == dict3[-vwp3a[[21823]]]))

## ---------------------------------------------------------------------
## H. MAPPING PROBE SET IDS BETWEEN CHIPS WITH vwhichPDict()
## ---------------------------------------------------------------------
## Here we show a simple (and very naive) algorithm for mapping probe
## set IDs between the hgu95av2 and hgu133a chips (Affymetrix).
## 2 probe set IDs are considered mapped iff they share at least one
## probe.
## WARNING: This example takes about 10 minutes to run.
if (interactive()) {

  library(hgu95av2probe)
  library(hgu133aprobe)
  probes1 &lt;- DNAStringSet(hgu95av2probe)
  probes2 &lt;- DNAStringSet(hgu133aprobe)
  pdict2 &lt;- PDict(probes2)

  ## Get the mapping from probes1 to probes2 (based on exact matching):
  map1to2 &lt;- vwhichPDict(pdict2, probes1) 

  ## The following helper function uses the probe level mapping to induce
  ## the mapping at the probe set IDs level (from hgu95av2 to hgu133a).
  ## To keep things simple, 2 probe set IDs are considered mapped iff
  ## each of them contains at least one probe mapped to one probe of
  ## the other:
  mapProbeSetIDs1to2 &lt;- function(psID)
    unique(hgu133aprobe$Probe.Set.Name[unlist(
      map1to2[hgu95av2probe$Probe.Set.Name == psID]
    )])

  ## Use the helper function to build the complete mapping:
  psIDs1 &lt;- unique(hgu95av2probe$Probe.Set.Name)
  mapPSIDs1to2 &lt;- lapply(psIDs1, mapProbeSetIDs1to2)  # about 3 min.
  names(mapPSIDs1to2) &lt;- psIDs1

  ## Do some basic stats:
  table(sapply(mapPSIDs1to2, length))

  ## [ADVANCED USERS ONLY]
  ## An alternative that is slightly faster is to put all the probes
  ## (hgu95av2 + hgu133a) in a single PDict object and then query its
  ## 'dups0' slot directly. This slot is a Dups object containing the
  ## mapping between duplicated patterns.
  ## Note that we can do this only because all the probes have the
  ## same length (25) and because we are doing exact matching:

  probes12 &lt;- DNAStringSet(c(hgu95av2probe$sequence, hgu133aprobe$sequence))
  pdict12 &lt;- PDict(probes12)
  dups0 &lt;- pdict12@dups0

  mapProbeSetIDs1to2alt &lt;- function(psID)
  {
    ii1 &lt;- unique(togroup(dups0, which(hgu95av2probe$Probe.Set.Name == psID)))
    ii2 &lt;- members(dups0, ii1) - length(probes1)
    ii2 &lt;- ii2[ii2 &gt;= 1L]
    unique(hgu133aprobe$Probe.Set.Name[ii2])
  }

  mapPSIDs1to2alt &lt;- lapply(psIDs1, mapProbeSetIDs1to2alt)  # about 5 min.
  names(mapPSIDs1to2alt) &lt;- psIDs1

  ## 'mapPSIDs1to2alt' and 'mapPSIDs1to2' contain the same mapping:
  stopifnot(identical(lapply(mapPSIDs1to2alt, sort),
                      lapply(mapPSIDs1to2, sort)))
}
</pre>

<hr /><div style="text-align: center;">[Package <em>Biostrings</em> version 2.50.1 <a href="00Index.html">Index</a>]</div>
</body></html>
