<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: (Deprecated) Needleman-Wunsch Global Alignment</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for needwunsQS {Biostrings}"><tr><td>needwunsQS {Biostrings}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>(Deprecated) Needleman-Wunsch Global Alignment</h2>

<h3>Description</h3>

<p>Simple gap implementation of Needleman-Wunsch global alignment algorithm.</p>


<h3>Usage</h3>

<pre>
needwunsQS(s1, s2, substmat, gappen = 8)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>s1, s2</code></td>
<td>
<p>an R character vector of length 1 or an <a href="XString-class.html">XString</a> object.</p>
</td></tr>
<tr valign="top"><td><code>substmat</code></td>
<td>
<p>matrix of alignment score values.</p>
</td></tr>
<tr valign="top"><td><code>gappen</code></td>
<td>
<p>penalty for introducing a gap in the alignment.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Follows specification of Durbin, Eddy, Krogh, Mitchison (1998).
This function has been deprecated and is being replaced by
<code>pairwiseAlignment</code>.
</p>


<h3>Value</h3>

<p>An instance of class <code>"PairwiseAlignments"</code>.
</p>


<h3>Author(s)</h3>

<p>Vince Carey (<a href="mailto:stvjc@channing.harvard.edu">stvjc@channing.harvard.edu</a>) (original author) and H. Pag√®s (current maintainer).</p>


<h3>References</h3>

<p>R. Durbin, S. Eddy, A. Krogh, G. Mitchison,
Biological Sequence Analysis, Cambridge UP 1998, sec 2.3.</p>


<h3>See Also</h3>

<p><a href="pairwiseAlignment.html">pairwiseAlignment</a>,
<a href="PairwiseAlignments-class.html">PairwiseAlignments-class</a>,
<a href="substitution_matrices.html">substitution.matrices</a>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
  ## This function has been deprecated
  ## Use 'pairwiseAlignment' instead.

  ## nucleotide alignment
  mat &lt;- matrix(-5L, nrow = 4, ncol = 4)
  for (i in seq_len(4)) mat[i, i] &lt;- 0L
  rownames(mat) &lt;- colnames(mat) &lt;- DNA_ALPHABET[1:4]
  s1 &lt;- DNAString(paste(sample(DNA_ALPHABET[1:4], 1000, replace=TRUE), collapse=""))
  s2 &lt;- DNAString(paste(sample(DNA_ALPHABET[1:4], 1000, replace=TRUE), collapse=""))
  nw0 &lt;- needwunsQS(s1, s2, mat, gappen = 0)
  nw1 &lt;- needwunsQS(s1, s2, mat, gappen = 1)
  nw5 &lt;- needwunsQS(s1, s2, mat, gappen = 5)

  ## amino acid alignment
  needwunsQS("PAWHEAE", "HEAGAWGHEE", substmat = "BLOSUM50")

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>Biostrings</em> version 2.50.1 <a href="00Index.html">Index</a>]</div>
</body></html>
