<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: XStringSet objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for XStringSet-class {Biostrings}"><tr><td>XStringSet-class {Biostrings}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>XStringSet objects</h2>

<h3>Description</h3>

<p>The BStringSet class is a container for storing a set of
<code><a href="XString-class.html">BString</a></code> objects and for making its manipulation
easy and efficient.
</p>
<p>Similarly, the DNAStringSet (or RNAStringSet, or AAStringSet) class is
a container for storing a set of <code><a href="DNAString-class.html">DNAString</a></code>
(or <code><a href="RNAString-class.html">RNAString</a></code>, or <code><a href="AAString-class.html">AAString</a></code>) objects.
</p>
<p>All those containers derive directly (and with no additional slots)
from the XStringSet virtual class.
</p>


<h3>Usage</h3>

<pre>
## Constructors:
BStringSet(x=character(), start=NA, end=NA, width=NA, use.names=TRUE)
DNAStringSet(x=character(), start=NA, end=NA, width=NA, use.names=TRUE)
RNAStringSet(x=character(), start=NA, end=NA, width=NA, use.names=TRUE)
AAStringSet(x=character(), start=NA, end=NA, width=NA, use.names=TRUE)

## Accessor-like methods:
## S4 method for signature 'character'
width(x)
## S4 method for signature 'XStringSet'
nchar(x, type="chars", allowNA=FALSE)

## ... and more (see below)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>Either a character vector (with no NAs), or an <a href="XString-class.html">XString</a>,
XStringSet or <a href="XStringViews-class.html">XStringViews</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>start,end,width</code></td>
<td>

<p>Either <code>NA</code>, a single integer, or an integer vector of the same
length as <code>x</code> specifying how <code>x</code> should be &quot;narrowed&quot;
(see <code>?<a href="../../IRanges/html/narrow.html">narrow</a></code> for the details).
</p>
</td></tr>
<tr valign="top"><td><code>use.names</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>. Should names be preserved?
</p>
</td></tr>
<tr valign="top"><td><code>type,allowNA</code></td>
<td>

<p>Ignored.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The <code>BStringSet</code>, <code>DNAStringSet</code>, <code>RNAStringSet</code> and
<code>AAStringSet</code> functions are constructors that can be used to
turn input <code>x</code> into an XStringSet object of the desired base type.
</p>
<p>They also allow the user to &quot;narrow&quot; the sequences contained in <code>x</code>
via proper use of the <code>start</code>, <code>end</code> and/or <code>width</code>
arguments. In this context, &quot;narrowing&quot; means dropping a prefix or/and
a suffix of each sequence in <code>x</code>.
The &quot;narrowing&quot; capabilities of these constructors can be illustrated
by the following property: if <code>x</code> is a character vector
(with no NAs), or an XStringSet (or <a href="XStringViews-class.html">XStringViews</a>) object,
then the 3 following transformations are equivalent:
</p>

<dl>
<dt></dt><dd>
<p><code>BStringSet(x, start=mystart, end=myend, width=mywidth)</code>
</p>
</dd>
<dt></dt><dd>
<p><code>subseq(BStringSet(x), start=mystart, end=myend, width=mywidth)</code>
</p>
</dd>
<dt></dt><dd>
<p><code>BStringSet(subseq(x, start=mystart, end=myend, width=mywidth))</code>
</p>
</dd>
</dl>

<p>Note that, besides being more convenient, the first form is also more
efficient on character vectors.
</p>


<h3>Accessor-like methods</h3>

<p>In the code snippets below,
<code>x</code> is an XStringSet object.
</p>

<dl>
<dt></dt><dd>
<p><code>length(x)</code>:
The number of sequences in <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>width(x)</code>:
A vector of non-negative integers containing the number
of letters for each element in <code>x</code>.
Note that <code>width(x)</code> is also defined for a character vector
with no NAs and is equivalent to <code>nchar(x, type="bytes")</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>names(x)</code>:
<code>NULL</code> or a character vector of the same length as <code>x</code>
containing a short user-provided description or comment for each
element in <code>x</code>.
These are the only data in an XStringSet object that can safely
be changed by the user. All the other data are immutable!
As a general recommendation, the user should never try to modify
an object by accessing its slots directly.
</p>
</dd>
<dt></dt><dd>
<p><code>alphabet(x)</code>:
Return <code>NULL</code>, <code><a href="DNAString-class.html">DNA_ALPHABET</a></code>,
<code><a href="RNAString-class.html">RNA_ALPHABET</a></code> or <code><a href="AAString-class.html">AA_ALPHABET</a></code> depending on
whether <code>x</code> is a BStringSet, DNAStringSet, RNAStringSet or
AAStringSet object.
</p>
</dd>
<dt></dt><dd>
<p><code>nchar(x)</code>:
The same as <code>width(x)</code>.
</p>
</dd>
</dl>



<h3>Subsequence extraction and related transformations</h3>

<p>In the code snippets below,
<code>x</code> is a character vector (with no NAs),
or an XStringSet (or <a href="XStringViews-class.html">XStringViews</a>) object.
</p>

<dl>
<dt></dt><dd>
<p><code>subseq(x, start=NA, end=NA, width=NA)</code>:
Applies <code>subseq</code> on each element in <code>x</code>.
See <code>?<a href="../../XVector/html/subseq.html">subseq</a></code> for the details.
</p>
<p>Note that this is similar to what <code><a href="../../base/html/substr.html">substr</a></code> does on a
character vector. However there are some noticeable differences:
</p>
<p>(1) the arguments are <code>start</code> and <code>stop</code> for
<code><a href="../../base/html/substr.html">substr</a></code>;
</p>
<p>(2) the SEW interface (start/end/width) interface of <code>subseq</code>
is richer (e.g. support for negative start or end values);
and (3) <code>subseq</code> checks that the specified start/end/width values
are valid i.e., unlike <code><a href="../../base/html/substr.html">substr</a></code>, it throws an error if
they define &quot;out of limits&quot; subsequences or subsequences with a
negative width.
</p>
</dd>
<dt></dt><dd>
<p><code>narrow(x, start=NA, end=NA, width=NA, use.names=TRUE)</code>:
Same as <code>subseq</code>. The only differences are: (1) <code>narrow</code>
has a <code>use.names</code> argument; and (2) all the things <code>narrow</code>
and <code>subseq</code> work on
(<a href="../../IRanges/html/IRanges.html">IRanges</a>, XStringSet or 
<a href="XStringViews-class.html">XStringViews</a> objects for <code>narrow</code>,
<a href="../../XVector/html/XVector.html">XVector</a> or XStringSet objects for
<code>subseq</code>). But they both work and do the same thing on an
XStringSet object. 
</p>
</dd>
<dt></dt><dd>
<p><code>threebands(x, start=NA, end=NA, width=NA)</code>:
Like the method for <a href="../../IRanges/html/IRanges.html">IRanges</a>
objects, the 
<code>threebands</code> methods for character vectors and XStringSet
objects extend the capability of <code>narrow</code> by returning the 3
set of subsequences (the left, middle and right subsequences)
associated to the narrowing operation.
See <code>?<a href="../../IRanges/html/threebands.html">threebands</a></code> in the
<span class="pkg">IRanges</span> package for the details.
</p>
</dd>
<dt></dt><dd>
<p><code>subseq(x, start=NA, end=NA, width=NA) &lt;- value</code>:
A vectorized version of the <code><a href="../../XVector/html/subseq%3C-.html">subseq&lt;-</a></code>
method for <a href="../../XVector/html/XVector.html">XVector</a> objects.
See <code>?`<a href="../../XVector/html/subseq%3C-.html">subseq&lt;-</a>`</code> for the details.
</p>
</dd>
</dl>



<h3>Subsetting and appending</h3>

<p>In the code snippets below,
<code>x</code> and <code>values</code> are XStringSet objects,
and <code>i</code> should be an index specifying the elements to extract.
</p>

<dl>
<dt></dt><dd>
<p><code>x[i]</code>:
Return a new XStringSet object made of the selected elements.
</p>
</dd>
<dt></dt><dd>
<p><code>x[[i]]</code>:
Extract the i-th <code><a href="XString-class.html">XString</a></code> object from <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>append(x, values, after=length(x))</code>:
Add sequences in <code>values</code> to <code>x</code>.
</p>
</dd>
</dl>



<h3>Set operations</h3>

<p>In the code snippets below,
<code>x</code> and <code>y</code> are XStringSet objects.
</p>

<dl>
<dt></dt><dd>
<p><code>union(x, y)</code>:
Union of <code>x</code> and <code>y</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>intersect(x, y)</code>:
Intersection of <code>x</code> and <code>y</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>setdiff(x, y)</code>:
Asymmetric set difference of <code>x</code> and <code>y</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>setequal(x, y)</code>:
Set equality of <code>x</code> to <code>y</code>.
</p>
</dd>
</dl>



<h3>Other methods</h3>

<p>In the code snippets below,
<code>x</code> is an XStringSet object.
</p>

<dl>
<dt></dt><dd>
<p><code>unlist(x)</code>:
Turns <code>x</code> into an <a href="XString-class.html">XString</a> object by combining the
sequences in <code>x</code> together.
Fast equivalent to <code>do.call(c, as.list(x))</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>as.character(x, use.names=TRUE)</code>:
Converts <code>x</code> to a character vector of the same length as <code>x</code>.
The <code>use.names</code> argument controls whether or not <code>names(x)</code>
should be propagated to the names of the returned vector.
</p>
</dd>
<dt></dt><dd>
<p><code>as.factor(x)</code>:
Converts <code>x</code> to a factor, via <code>as.character(x)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>as.matrix(x, use.names=TRUE)</code>:
Returns a character matrix containing the &quot;exploded&quot; representation of
the strings. Can only be used on an XStringSet object with
equal-width strings.
The <code>use.names</code> argument controls whether or not <code>names(x)</code>
should be propagated to the row names of the returned matrix.
</p>
</dd>
<dt></dt><dd>
<p><code>toString(x)</code>:
Equivalent to <code>toString(as.character(x))</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>show(x)</code>:
By default the <code>show</code> method displays 5 head and 5 tail
lines. The number of lines can be altered by setting the global
options <code>showHeadLines</code> and <code>showTailLines</code>. If the 
object length is less than the sum of the options, the full object 
is displayed. These options affect GRanges, GAlignments, IRanges,
and XStringSet objects.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>H. Pagès</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="XStringSet-io.html">readDNAStringSet</a></code> and <code><a href="XStringSet-io.html">writeXStringSet</a></code>
for reading/writing a <a href="XStringSet-class.html">DNAStringSet</a> object (or other
<a href="XStringSet-class.html">XStringSet</a> derivative) from/to a FASTA or FASTQ file.
</p>
</li>
<li> <p><a href="XStringSet-comparison.html">XStringSet-comparison</a>
</p>
</li>
<li> <p><a href="XString-class.html">XString</a> objects.
</p>
</li>
<li> <p><a href="XStringViews-class.html">XStringViews</a> objects.
</p>
</li>
<li> <p><a href="XStringSetList-class.html">XStringSetList</a> objects.
</p>
</li>
<li> <p><code><a href="../../XVector/html/subseq.html">subseq</a></code>, <code><a href="../../IRanges/html/narrow.html">narrow</a></code>,
and <code><a href="../../base/html/substr.html">substr</a></code>.
</p>
</li>
<li> <p><code><a href="../../XVector/html/compact.html">compact</a></code>
</p>
</li>
<li> <p><a href="../../XVector/html/XVectorList.html">XVectorList</a> objects.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. USING THE XStringSet CONSTRUCTORS ON A CHARACTER VECTOR OR FACTOR
## ---------------------------------------------------------------------
## Note that there is no XStringSet() constructor, but an XStringSet
## family of constructors: BStringSet(), DNAStringSet(), RNAStringSet(),
## etc...
x0 &lt;- c("#CTC-NACCAGTAT", "#TTGA", "TACCTAGAG")
width(x0)
x1 &lt;- BStringSet(x0)
x1

## 3 equivalent ways to obtain the same BStringSet object:
BStringSet(x0, start=4, end=-3)
subseq(x1, start=4, end=-3)
BStringSet(subseq(x0, start=4, end=-3))

dna0 &lt;- DNAStringSet(x0, start=4, end=-3)
dna0
names(dna0)
names(dna0)[2] &lt;- "seqB"
dna0

## When the input vector contains a lot of duplicates, turning it into
## a factor first before passing it to the constructor will produce an
## XStringSet object that is more compact in memory:
library(hgu95av2probe)
x2 &lt;- sample(hgu95av2probe$sequence, 999000, replace=TRUE)
dna2a &lt;- DNAStringSet(x2)
dna2b &lt;- DNAStringSet(factor(x2))  # slower but result is more compact
object.size(dna2a)
object.size(dna2b)

## ---------------------------------------------------------------------
## B. USING THE XStringSet CONSTRUCTORS ON A SINGLE SEQUENCE (XString
##    OBJECT OR CHARACTER STRING)
## ---------------------------------------------------------------------
x3 &lt;- "abcdefghij"
BStringSet(x3, start=2, end=6:2)  # behaves like 'substring(x3, 2, 6:2)'
BStringSet(x3, start=-(1:6))
x4 &lt;- BString(x3)
BStringSet(x4, end=-(1:6), width=3)

## Randomly extract 1 million 40-mers from C. elegans chrI:
extractRandomReads &lt;- function(subject, nread, readlength)
{
    if (!is.integer(readlength))
        readlength &lt;- as.integer(readlength)
    start &lt;- sample(length(subject) - readlength + 1L, nread,
                    replace=TRUE)
    DNAStringSet(subject, start=start, width=readlength)
}
library(BSgenome.Celegans.UCSC.ce2)
rndreads &lt;- extractRandomReads(Celegans$chrI, 1000000, 40)
## Notes:
## - This takes only 2 or 3 seconds versus several hours for a solution
##   using substring() on a standard character string.
## - The short sequences in 'rndreads' can be seen as the result of a
##   simulated high-throughput sequencing experiment. A non-realistic
##   one though because:
##     (a) It assumes that the underlying technology is perfect (the
##         generated reads have no technology induced errors).
##     (b) It assumes that the sequenced genome is exactly the same as the
##         reference genome.
##     (c) The simulated reads can contain IUPAC ambiguity letters only
##         because the reference genome contains them. In a real
##         high-throughput sequencing experiment, the sequenced genome
##         of course doesn't contain those letters, but the sequencer
##         can introduce them in the generated reads to indicate ambiguous
##         base-calling.
##     (d) The simulated reads come from the plus strand only of a single
##         chromosome.
## - See the getSeq() function in the BSgenome package for how to
##   circumvent (d) i.e. how to generate reads that come from the whole
##   genome (plus and minus strands of all chromosomes).

## ---------------------------------------------------------------------
## C. USING THE XStringSet CONSTRUCTORS ON AN XStringSet OBJECT
## ---------------------------------------------------------------------
library(drosophila2probe)
probes &lt;- DNAStringSet(drosophila2probe)
probes

RNAStringSet(probes, start=2, end=-5)  # does NOT copy the sequence data!

## ---------------------------------------------------------------------
## D. USING THE XStringSet CONSTRUCTORS ON AN ORDINARY list OF XString
##    OBJECTS
## ---------------------------------------------------------------------
probes10 &lt;- head(probes, n=10)
set.seed(33)
shuffled_nucleotides &lt;- lapply(probes10, sample)
shuffled_nucleotides

DNAStringSet(shuffled_nucleotides)  # does NOT copy the sequence data!

## Note that the same result can be obtained in a more compact way with
## just:
set.seed(33)
endoapply(probes10, sample)

## ---------------------------------------------------------------------
## E. USING subseq() ON AN XStringSet OBJECT
## ---------------------------------------------------------------------
subseq(probes, start=2, end=-5)

subseq(probes, start=13, end=13) &lt;- "N"
probes

## Add/remove a prefix:
subseq(probes, start=1, end=0) &lt;- "--"
probes
subseq(probes, end=2) &lt;- ""
probes

## Do more complicated things:
subseq(probes, start=4:7, end=7) &lt;- c("YYYY", "YYY", "YY", "Y")
subseq(probes, start=4, end=6) &lt;- subseq(probes, start=-2:-5)
probes

## ---------------------------------------------------------------------
## F. UNLISTING AN XStringSet OBJECT
## ---------------------------------------------------------------------
library(drosophila2probe)
probes &lt;- DNAStringSet(drosophila2probe)
unlist(probes)

## ---------------------------------------------------------------------
## G. COMPACTING AN XStringSet OBJECT
## ---------------------------------------------------------------------
## As a particular type of XVectorList objects, XStringSet objects can
## optionally be compacted. Compacting is done typically before
## serialization. See ?compact for more information.
library(drosophila2probe)
probes &lt;- DNAStringSet(drosophila2probe)

y &lt;- subseq(probes[1:12], start=5)
probes@pool
y@pool
object.size(probes)
object.size(y)

y0 &lt;- compact(y)
y0@pool
object.size(y0)
</pre>

<hr /><div style="text-align: center;">[Package <em>Biostrings</em> version 2.50.1 <a href="00Index.html">Index</a>]</div>
</body></html>
